c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,457 @@
+        const auto it = canonicalCombiningClass_.find(c);
+        if (it == canonicalCombiningClass_.end()) {
+            return 0;
+        } else {
+            return it->second;
+        }
+    }
+
+    std::optional<std::uint64_t> naturalNumericValue(const char32_t c) const {
+        const auto it = naturalNumericValue_.find(c);
+        if (it == naturalNumericValue_.end()) {
+            return std::nullopt;
+        } else {
+            return it->second;
+        }
+    }
+
+    const CodePointSet& codePointsWithNumericValue() const {
+        return codePointsWithNumericValue_;
+    }
+
+   private:
+    void readMultiPropertyFile(const std::filesystem::path& file) {
+        std::ifstream stream(file);
+        CHECK(stream.good());
+        std::cout << "Reading " << file << "...\n";
+        for (std::string line; std::getline(stream, line);) {
+            // See https://www.unicode.org/reports/tr44/#UnicodeData.txt.
+            const auto record = stripComment(line);
+            if (record.empty()) {
+                continue;
+            }
+            switch (std::count(record.begin(), record.end(), ';')) {
+                case 1: {
+                    const auto [cp, propertyName] = fields<2>(record);
+                    const CodePointRange range = parseHexCodePointRange(cp);
+                    binaryProperties_[std::string(propertyName)].addAll(range);
+                    break;
+                }
+                case 2:  // Enumerated property.
+                    break;
+                default:
+                    std::terminate();
+            }
+        }
+    }
+
+    // Keys are C, L, M, N, P, S, Z.
+    std::map<char, CodePointSet> coarseGeneralCategory_;
+    // Keyed on short alias.
+    std::map<std::string, CodePointSet, std::less<>> generalCategory_;
+    std::map<std::string, CodePointSet, std::less<>> binaryProperties_;
+    // Default <code point> omitted.
+    std::map<char32_t, char32_t> simpleUppercaseMapping_;
+    // Default <code point> omitted.
+    std::map<char32_t, char32_t> simpleLowercaseMapping_;
+    // Default 0 omitted.
+    std::map<char32_t, std::uint8_t> canonicalCombiningClass_;
+    std::map<char32_t, std::uint64_t> naturalNumericValue_;
+    // Characters with a non-NaN Numeric_Value.
+    CodePointSet codePointsWithNumericValue_;
+};
+
+UCD* const ucd = new UCD();
+
+static bool isSifterNonSpacingCombining(char32_t c) {
+    // The DUCET Sifter is using these PUA code points in artificial
+    // decompositions. They are treated like gc=Mn, so that they get secondary
+    // weights.
+    return 0xF8F0 <= c && c <= 0xF8FF;
+}
+
+static bool isSpecialCJKIdeograph(char32_t c) {
+    return 0xFA0E <= c && c <= 0xFA29 &&
+           (c <= 0xFA0F || c == 0xFA11 || c == 0xFA13 || c == 0xFA14 ||
+            c == 0xFA1F || c == 0xFA21 || c == 0xFA23 || c == 0xFA24 ||
+            0xFA27 <= c);
+}
+
+}  // namespace
+
+extern "C" int unisift_IsAlphabetic(std::uint32_t c) {
```

Comment:
nit: I have never seen uint32_t qualified with `std::` ...

Reply:
It depends whether you get them from the C++ header or from the C header; I’m doing it the C++ way.